### ISSUE TITLE
Make sure build is completed before `make test` starts tests

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#22943]
Copying @hlandau 's description (with editorial adaptations) from [anther issue](https://github.com/openssl/openssl/issues/21999#issuecomment-1821380568):

When developing I always find running `make test` after changing a file to result in trying to run some test binaries 
before they are actually finished compiling, resulting in an error, which is why I always run `make && make test`.
We should fix that.

**Comment by levitte** [comment]
If you do just `make test`, I don't see how that could happen.  However, if you somehow had `-j{n}` (explicit or implied somehow), then you're outta luck, because of the parallel dependency where the last one is `run-tests`:

https://github.com/openssl/openssl/blob/7ebaab7689f66ede1f960c42be3446922e3f5e21/Configurations/unix-Makefile.tmpl#L560-L563

**Comment by DDvO** [comment]
Yep, this must bei the point: parallel `make` goes wrong then.

**Comment by rwfranks** [comment]
The real issue here lies in the declared dependency relationships.
```
##@ Testing
test: tests ## Run tests (alias of "tests")
tests: build_generated ## Run tests
        $(MAKE) depend && $(MAKE) _tests
_tests: build_programs_nodep build_modules_nodep link-utils run_tests
run_tests:
```
The way things stand, from make's pov  "build_programs_nodep", "build_modules_nodep", "link-utils" and "run_tests" are declared to be mutually order independent (i.e. potentially parallel jobs).  Neither do I believe that make offers any guarantee to resolve target lists left to right, so in theory even make -j1 could fail.
The build stuff needs to be grouped under a separate target and the tests depend on that target.

**Comment by slontis** [comment]
Recently my 'make test' stopped working.. I am not sure what I have done.. (listing the tests shows nothing also)
Adding the following manually to the Makefile addresses the issue..
.PHONY: run_tests

**Comment by levitte** [comment]
I hope I fixed it.  Please have a look at the PRs

**Comment by rwfranks** [comment]
PRs look ok.

**Comment by levitte** [comment]
Reopening, 'cause #22948 hasn't been merged yet

**Comment by jamuir** [comment]
@levitte :  can this be closed now?

**Comment by DDvO** [comment]
Closing as this has been merged to 3.0, 3.1, 3.2, and `master`.
